sdk/cosmos/azcosmos/async_cache.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"context"
+	"reflect"
+	"sync"
+)
+
+type invalidCacheValue struct{}
+
+func (i invalidCacheValue) Error() string { return "Invalid cache value" }
+
+type asyncCache struct {
+	values sync.Map
+}
+
+type cacheValueTask func() interface{}
+
+type cacheValue struct {
+	value    interface{}
+	complete bool
+	fn       cacheValueTask
+	ch       <-chan interface{}
+}
+
+func newAsyncCache() *asyncCache {
+	return &asyncCache{}
+}
+
+func (ac *asyncCache) set(key interface{}, value interface{}) {
+	ac.values.Store(key, cacheValue{value: value})
+}
+
+func (ac *asyncCache) setAsync(key interface{}, singleValueInit cacheValueTask, ctx context.Context) error {
+	ch := ac.execCacheValueTask(singleValueInit)
```

Comment:
How does the flow work on concurrent calls to setAsync? Wouldn't both create a channel and both call Store?

Reply:
set is synchronous and changed the name to reflect this.  set is used to initialize the cache. getAsync is used to asynchronously get the value from the cache and update the value depending on the obsolete value passed.